src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,477 @@
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> covs(
+      num_params, num_params);
+  covariance_computer.GetCovarianceMatrixInTangentSpace(parameter_blocks,
+                                                        covs.data());
+  cov_variables_ = covs;
+  cov_poses_ = cov_variables_.block(0, 0, num_params_poses_, num_params_poses_);
+  return true;
+}
+
+bool BundleAdjustmentCovarianceEstimatorCeresBackend::Compute() {
+  ceres::Covariance::Options options;
+  ceres::Covariance covariance_computer(options);
+  std::vector<const double*> parameter_blocks;
+  for (const auto& block : pose_blocks_) {
+    parameter_blocks.push_back(block);
+  }
+  try {
+    covariance_computer.Compute(parameter_blocks, problem_);
+  } catch (const std::exception& e) {
+    return false;
+  }
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> covs(
+      num_params_poses_, num_params_poses_);
+  covariance_computer.GetCovarianceMatrixInTangentSpace(parameter_blocks,
+                                                        covs.data());
+  cov_poses_ = covs;
+  return true;
+}
+
+void BundleAdjustmentCovarianceEstimator::SetLambda(double lambda) {
+  lambda_ = lambda;
+}
+
+void BundleAdjustmentCovarianceEstimator::ComputeSchurComplement() {
+  // Evaluate jacobian
+  LOG(INFO) << "Evaluate jacobian matrix";
+  ceres::Problem::EvaluateOptions eval_options;
+  eval_options.parameter_blocks.clear();
+  for (const auto& block : pose_blocks_) {
+    eval_options.parameter_blocks.push_back(const_cast<double*>(block));
+  }
+  for (const auto& block : other_variables_blocks_) {
+    eval_options.parameter_blocks.push_back(const_cast<double*>(block));
+  }
+  for (const auto& block : point_blocks_) {
+    eval_options.parameter_blocks.push_back(const_cast<double*>(block));
+  }
+  ceres::CRSMatrix J_full_crs;
+  problem_->Evaluate(eval_options, nullptr, nullptr, nullptr, &J_full_crs);
+  int num_residuals = J_full_crs.num_rows;
+  int num_params = J_full_crs.num_cols;
+  Eigen::Map<Eigen::SparseMatrix<double, Eigen::RowMajor>> J_full(
+      J_full_crs.num_rows,
+      J_full_crs.num_cols,
+      J_full_crs.values.size(),
+      J_full_crs.rows.data(),
+      J_full_crs.cols.data(),
+      J_full_crs.values.data());
+
+  // Step 3: Schur elimination on points
+  LOG(INFO) << "Schur elimination on points";
+  Eigen::SparseMatrix<double> J_c =
+      J_full.block(0, 0, num_residuals, num_params - num_params_points_);
+  Eigen::SparseMatrix<double> J_p = J_full.block(
+      0, num_params - num_params_points_, num_residuals, num_params_points_);
+  Eigen::SparseMatrix<double> H_cc = J_c.transpose() * J_c;
+  Eigen::SparseMatrix<double> H_cp = J_c.transpose() * J_p;
+  Eigen::SparseMatrix<double> H_pc = H_cp.transpose();
+  Eigen::SparseMatrix<double> H_pp = J_p.transpose() * J_p;
+  // in-place computation of H_pp_inv
+  Eigen::SparseMatrix<double> H_pp_inv = H_pp;  // shallow copy
```

Comment:
Maybe I misunderstand how Eigen works but isn't this a full copy of the data?

Reply:
Yes. I changed this into reference now.